### PR TITLE
explore/10-exploration-strategy

### LINKED ISSUE #10: hintset exploration strategy
# Task

Investigate possible learning algorithms (i.e. strategies of exploration for good transition).

# Context

1. We have realized that the robustness problem is quite acute even on commonly used benchmarks. 
2. The natural way to deal with it would be to a) switch to offline learning and b) use checks of similarity of the custom plan and its estimated cardinalities with the experience from history. 
3. In order to guarantee the safety of any prediction of the `M` model, the transitions obtained with it **must have already been explored**. In this case, we **don't need any prediction model**, because we can just take the times from history itself! 
4. It means that offline learning becomes just applying a *smart strategy for filling history with the most useful transitions*, i.e., we must explore queries and hintsets in such a way that we can find transitions with the highest speedup as quickly as possible. 
5. So we get a situation where hintsets **are just a way to get the desired transition**, and inference becomes **just a search** against the default plan for hintsets that could potentially lead to good and already confirmed transition. We will see later why this is an extremely important feature of the model.

### PR DESCRIPTION
**Pull request by zinchse**
**tldr;** a general scheme of the local search algorithm was proposed; in practice, it was possible to show that the algorithm is able to quickly reach the suboptimal by avoiding exploration of the entire solution space; special techniques were also developed to reduce the training time;

## Key Insights:
- the degree of parallelism is an **extremely** important feature
- it is not necessary to turn off operations to get the fastest possible result - one iteration of local search with aggressive timeout, in which only the degree of parallelism and the join operator are changed, is enough

## Results for `JOB`

| `SearchingSetting` | Boost (% of optimal) | Learning Time (s) |
|-|-|-|
| `DISABLE_OP` | 69.2 % | 10149.92 s |
| `FORCE_JOIN` | 52.2 % | **1911.83 s** |
| `DISABLE_OP` OR `DECREASE_DOP` | 78.1 % | 13114.70 s |
| (`FORCE_JOIN` OR `DISABLE_INL`  OR `DECREASE_DOP`) | **87.9 %** | **7374.47 s** |
| (`FORCE_JOIN` OR `DISABLE_INL` OR `DISABLE_OP` OR `DECREASE_DOP`) | 91.5 % | 17072.23 s |
| (`FORCE_JOIN` OR `DISABLE_INL`) AND `CHANGE_DOP` | **94.2 %** | **12112.15 s** |
| (`FORCE_JOIN` OR `DISABLE_INL` OR `DISABLE_OP` OR `DECREASE_DOP`) AND `CHANGE_DOP` | 98.5 % | 37748.68 s |

### CODE REVIEW

**Comment by zinchse** [comment]
great;  there is no need to merge this into the main branch; I will take some functions by copy-paste or checkout of individual files;

all that remains is to:
1) test the hypothesis of using NN to rank neighbors;
2) emulate the E2E approach with saving transitions and inference through planner calls
